File: s3/PHP/lib/s3client.php
<?php
declare(strict_types=1);
/*
 * Copyright Core DF
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * S3-compatible object storage helpers (AWS S3, MinIO, etc.).
 */
require_once __DIR__ . '/../../../http/PHP/lib/result.php';

$autoload = __DIR__ . '/../vendor/autoload.php';
if (is_file($autoload)) {
    require_once $autoload;
}

use Aws\S3\S3Client as AwsS3Client;

final class S3client
{
    private static function region(): string
    {
        $r = getenv('AWS_REGION') ?: getenv('AWS_DEFAULT_REGION') ?: '';
        return $r !== '' && $r !== false ? $r : 'us-east-1';
    }

    private static function aws(): AwsS3Client
    {
        $config = ['version' => 'latest', 'region' => self::region()];
        $endpoint = getenv('S3_ENDPOINT_URL') ?: '';
        if ($endpoint !== '') {
            $config['endpoint'] = $endpoint;
            $config['use_path_style_endpoint'] = true;
        }
        return new AwsS3Client($config);
    }

    private static function bucket(?string $explicit): string
    {
        if ($explicit !== null && $explicit !== '') {
            return $explicit;
        }
        $b = getenv('S3_BUCKET');
        return $b !== false ? $b : '';
    }

    public static function Init(): array
    {
        $key = getenv('AWS_ACCESS_KEY_ID') ?: '';
        $profile = getenv('AWS_PROFILE') ?: '';
        if ($key === '' && $profile === '') {
            return CoreautoResult::missingEnv('AWS_ACCESS_KEY_ID/AWS_SECRET_ACCESS_KEY or AWS_PROFILE');
        }
        if (self::bucket(null) === '') {
            return CoreautoResult::missingEnv('S3_BUCKET (or pass bucket per call)');
        }
        return ['status_code' => 200];
    }

    public static function GetObject(string $key, ?string $bucketName = null): array
    {
        $b = self::bucket($bucketName);
        if ($b === '') {
            return CoreautoResult::missingEnv('S3_BUCKET');
        }
        if (!class_exists(AwsS3Client::class)) {
            return CoreautoResult::transportError('Run composer install in s3/PHP (aws/aws-sdk-php)');
        }
        try {
            $result = self::aws()->getObject(['Bucket' => $b, 'Key' => $key]);
            $content = (string) ($result['Body'] ?? '');
            return ['status_code' => 200, 'content' => $content];
        } catch (Throwable $e) {
            return CoreautoResult::transportError($e->getMessage());
        }
    }

    public static function PutObject(string $key, string $content, ?string $bucketName = null): array
    {
        $b = self::bucket($bucketName);
        if ($b === '') {
            return CoreautoResult::missingEnv('S3_BUCKET');
        }
        if (!class_exists(AwsS3Client::class)) {
            return CoreautoResult::transportError('Run composer install in s3/PHP (aws/aws-sdk-php)');
        }
        try {
            self::aws()->putObject([
                'Bucket' => $b,
                'Key' => $key,
                'Body' => $content,
            ]);
            return ['status_code' => 200];
        } catch (Throwable $e) {
            return CoreautoResult::transportError($e->getMessage());
        }
    }

    public static function ListObjects(string $prefix = '', ?string $bucketName = null): array
    {
        $b = self::bucket($bucketName);
        if ($b === '') {
            return CoreautoResult::missingEnv('S3_BUCKET');
        }
        if (!class_exists(AwsS3Client::class)) {
            return CoreautoResult::transportError('Run composer install in s3/PHP (aws/aws-sdk-php)');
        }
        try {
            $result = self::aws()->listObjectsV2([
                'Bucket' => $b,
                'Prefix' => $prefix,
            ]);
            $keys = [];
            foreach ($result['Contents'] ?? [] as $obj) {
                if (isset($obj['Key'])) {
                    $keys[] = $obj['Key'];
                }
            }
            return ['status_code' => 200, 'keys' => $keys];
        } catch (Throwable $e) {
            return CoreautoResult::transportError($e->getMessage());
        }
    }
}

File: s3/DotNet/S3Client.cs
// Copyright Core DF
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using Amazon;
using Amazon.S3;
using Amazon.S3.Model;

namespace CoreAuto.S3;

public static class S3Client
{
    public static Result Init()
    {
        if (string.IsNullOrEmpty(Env("AWS_ACCESS_KEY_ID")) && string.IsNullOrEmpty(Env("AWS_PROFILE")))
            return Result.MissingEnv("AWS_ACCESS_KEY_ID/AWS_SECRET_ACCESS_KEY or AWS_PROFILE");
        if (string.IsNullOrEmpty(Env("S3_BUCKET"))) return Result.MissingEnv("S3_BUCKET (or pass bucket per call)");
        return Result.Ok();
    }

    public static async Task<Result> GetObjectAsync(string key, string? bucket = null)
    {
        var b = Bucket(bucket); if (string.IsNullOrEmpty(b)) return Result.MissingEnv("S3_BUCKET");
        try {
            using var c = Client();
            var resp = await c.GetObjectAsync(b, key);
            using var reader = new StreamReader(resp.ResponseStream);
            return Result.Ok(new() { ["content"] = await reader.ReadToEndAsync() });
        } catch (Exception ex) { return Result.TransportError(ex.Message); }
    }

    public static async Task<Result> PutObjectAsync(string key, string content, string? bucket = null)
    {
        var b = Bucket(bucket); if (string.IsNullOrEmpty(b)) return Result.MissingEnv("S3_BUCKET");
        try { using var c = Client(); await c.PutObjectAsync(new PutObjectRequest { BucketName = b, Key = key, ContentBody = content }); return Result.Ok(); }
        catch (Exception ex) { return Result.TransportError(ex.Message); }
    }

    public static async Task<Result> ListObjectsAsync(string prefix = "", string? bucket = null)
    {
        var b = Bucket(bucket); if (string.IsNullOrEmpty(b)) return Result.MissingEnv("S3_BUCKET");
        try {
            using var c = Client();
            var resp = await c.ListObjectsV2Async(new ListObjectsV2Request { BucketName = b, Prefix = prefix });
            return Result.Ok(new() { ["keys"] = resp.S3Objects.Select(o => o.Key).ToList() });
        } catch (Exception ex) { return Result.TransportError(ex.Message); }
    }

    private static AmazonS3Client Client()
    {
        var cfg = new AmazonS3Config { RegionEndpoint = RegionEndpoint.GetBySystemName(Env("AWS_REGION", Env("AWS_DEFAULT_REGION", "us-east-1"))) };
        var ep = Env("S3_ENDPOINT_URL"); if (!string.IsNullOrEmpty(ep)) cfg.ServiceURL = ep;
        return new AmazonS3Client(cfg);
    }

    private static string Bucket(string? explicitBucket) => string.IsNullOrEmpty(explicitBucket) ? Env("S3_BUCKET") : explicitBucket;
    private static string Env(string k, string d = "") => Environment.GetEnvironmentVariable(k) ?? d;
}
